src/application-manager.coffee

```diff
@@ -0,0 +1,875 @@
+		dependenciesMetForKill = =>
+			!needsDownload and @_dependenciesMetForServiceKill(target, targetApp, availableImages)
+
+		# If the service is using a network or volume that is being updated, we need to kill it
+		# even if its strategy is handover
+		needsSpecialKill = @_hasCurrentNetworksOrVolumes(current, networkPairs, volumePairs)
+
+		if current?.isSameContainer(target)
+			# We're only stopping/starting it
+			return @_updateContainerStep(current, target)
+		else if !current?
+			# Either this is a new service, or the current one has already been killed
+			return @_fetchOrStartStep(current, target, needsDownload, dependenciesMetForStart)
+		else
+			strategy = checkString(target.labels['io.resin.update.strategy'])
+			validStrategies = [ 'download-then-kill', 'kill-then-download', 'delete-then-download', 'hand-over' ]
+			strategy = 'download-then-kill' if !_.includes(validStrategies, strategy)
+			timeout = checkInt(target.labels['io.resin.update.handover_timeout'])
+			return @_strategySteps[strategy](current, target, needsDownload, dependenciesMetForStart, dependenciesMetForKill, needsSpecialKill, timeout)
+
+	_nextStepsForAppUpdate: (currentApp, targetApp, availableImages = [], stepsInProgress = []) =>
+		emptyApp = { services: [], volumes: {}, networks: {} }
+		if !targetApp?
+			targetApp = emptyApp
+		if !currentApp?
+			currentApp = emptyApp
+		appId = targetApp.appId ? currentApp.appId
+		# Create the default network for the target app
+		targetApp.networks[targetApp.appId] ?= {}
+		Promise.join(
+			@compareNetworksOrVolumesForUpdate(@networks, { current: currentApp.networks, target: targetApp.networks }, appId)
+			@compareNetworksOrVolumesForUpdate(@volumes, { current: currentApp.volumes, target: targetApp.volumes }, appId)
+			@compareServicesForUpdate(currentApp.services, targetApp.services)
+			@compareImagesForMetadataUpdate(availableImages, targetApp.services)
+			(networkPairs, volumePairs, { removePairs, installPairs, updatePairs }, imagePairs) =>
+				steps = []
+				# All removePairs get a 'kill' action
+				_.forEach removePairs, ({ current }) ->
+					steps.push(serviceAction('kill', current.serviceId, current, null))
+				# next step for install pairs in download - start order, but start requires dependencies, networks and volumes met
+				# next step for update pairs in order by update strategy. start requires dependencies, networks and volumes met.
+				_.forEach installPairs.concat(updatePairs), (pair) =>
+					step = @_nextStepForService(pair, { targetApp, networkPairs, volumePairs, installPairs, updatePairs, stepsInProgress, availableImages })
+					steps.push(step) if step?
+				# next step for network pairs - remove requires services killed, create kill if no pairs or steps affect that service
+				_.forEach networkPairs, (pair) =>
+					pairSteps = @_nextStepsForNetwork(pair, currentApp, removePairs.concat(updatePairs))
+					steps = steps.concat(pairSteps) if !_.isEmpty(pairSteps)
```

Comment:
Why do we need to check if the `pairSteps` are empty? Concatenating an empty array is a no-op so you don't need the special case

Reply:
Done